### PR TITLE
Multiple mapping and matching

### PR DESCRIPTION
**Pull request by RomanChernetskyi**
## Purpose
Implement multiple mapping and matching

### MODDICORE-340
A new Factory was created for the new Multiple Entities in the scope of [MODDICORE-336](https://issues.folio.org/browse/MODDICORE-336). The mechanism for Multiple Holdings will create a new Holdings-entity for each UNIQUE location (multiple fields for holdings should be placed in the same field with permLocId) in the MARC file.
### MODDICORE-90
The mechanism for Multiple Holdings will create a new Item-entity for each marc field that defines holdings.permanentLocationId
### MODDICORE-317
Added support for multiple matching, setting up ERRORS and amount of non-matched holdings/items in context

### CODE REVIEW

**Review by KaterynaSenchenko — APPROVED** [review]

File: ramls/settings/authoritysourcefile.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     },
     "codes": {
       "type" : "array",
+      "description": "array of codes",
```
Comment (VRohach) [inline ramls/settings/authoritysourcefile.json:16]: @RomanChernetskyi should we add these properties manually? Can we just update raml-storage?

**Review by VRohach — APPROVED** [review]

File: src/main/java/org/folio/processing/mapping/mapper/mappers/HoldingsMapper.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    Optional<MappingRule> permanentLocationMappingRule = mappingRules.stream().filter(rule -> rule.getName().equals(PERMANENT_LOCATION_ID)).findFirst();
+
+    if (permanentLocationMappingRule.isEmpty() || !isStaredWithMarcField(permanentLocationMappingRule.get().getValue())) {
+      adjustContextToContainEntitiesAsJsonObject(eventPayload, EntityType.HOLDINGS);
+      writer.initialize(eventPayload);
+      holdings.add(mapSingleEntity(eventPayload, reader, writer, mappingRules, HOLDINGS));
+    } else {
+      String expressionPart = permanentLocationMappingRule.get().getValue().split(WHITESPACE_DIVIDER)[0];
+      String marcField = retrieveMarcFieldName(expressionPart)
+        .orElseThrow(() -> new RuntimeException(String.format("Invalid  value for mapping rule: %s", PERMANENT_LOCATION_ID)));
```
Comment (RuslanLavrov) [inline src/main/java/org/folio/processing/mapping/mapper/mappers/HoldingsMapper.java:70]:
```suggestion
        .orElseThrow(() -> new MappingException(String.format("Invalid  value for mapping rule: %s", PERMANENT_LOCATION_ID)));
```

Comment (RomanChernetskyi) [reply src/main/java/org/folio/processing/mapping/mapper/mappers/HoldingsMapper.java]: Done, thanks

File: ramls/settings/authoritysourcefile.json (hunk 1 above)
Comment (VRohach) [reply ramls/settings/authoritysourcefile.json:16]: Discussed with Roman - everything ok. We need to update it manually. Resolve conversation.

Comment (RomanChernetskyi) [reply ramls/settings/authoritysourcefile.json:16]: Hi, we should add these properties manually, because these files are not from raml-storage

**Pull request merged**